WHY.md

```diff
@@ -0,0 +1,107 @@
+# Why we wrote Beancounter
+### _and why is wasn't trivial_
+Intuitively, computing the balance for a Bitcoin wallet should be easy. This document explains the issues we ran into with both, our hot and cold wallets. We solved these issues by implementing Beancounter, a tool we then open-sourced.
+
+# How is the blockchain structured?
+A simplified explanation: Bitcoin stores transactions on a blockchain. A blockchain is a list of blocks. Each block contains transactions.
+
+A transaction consumes previous transaction:index pairs (debits) and creates new outputs (credits). The outputs are address:amount pairs and are called unspent transactions (UTXO) until they get consumed in a subsequent transaction.
+
+```
+            [...] -> [block dc52a] -> [...]
+                           |
+                           +-- transaction d1e6d
+                           | +-- debit: prevtx 068aa:1
+                           | +-- debit: prevtx bfa74:4
+                           | +-- credit: some address:amount
+                           | +-- credit: another address:amount
+                           | ...
+                           |
+                           +- transaction 1c867
+                           | ...
+                           |
+                           ...
+```
+
+The important point for our discussion is that the funds are the UTXOs and there’s no per-address state. Each Bitcoin client needs to compute the per-address balance for the wallets they care about.
+
+_Note: not all cryptocurrencies are implemented in this way. For instance, Ethereum maintains contract state._
+
+https://github.com/bitcoinbook/bitcoinbook (or the paper version of the book) is a good place to learn more about the underlying structures.
+
+# What is a Bitcoin wallet?
+A wallet is a collection of addresses. For [Hierarchical Deterministic wallets][bip32] (which is what we use), each wallet has a huge number of addresses (2^31). To figure out the balance, we need to find each addresses’ balance and sum things up. Doing the following would however be extremely slow:
```

Comment:
don't we have multiple levels in a path, so it's actually even bigger?  Or does beancounter not handle that?

Reply:
typically, a given derivation is fixed, except for the 2^31 addresses * 2 (change vs non-change). So 2^32.